File: 20.valid-parentheses.java
import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */
class Solution {
    public boolean isValid(String s) {
        if (s.length() == 0)
            return true;
        if (s.length() % 2 != 0)
            return false;
        Deque<Character> st = new LinkedList<>();
        Map<Character, Character> map = new HashMap<>();
        map.put('}', '{');
        map.put(']', '[');
        map.put(')', '(');
        for (int i = 0; i < s.length(); i++) {
            Character ch = new Character(s.charAt(i));
            Character pre = st.peek();
            Character tmp = map.get(ch);
            if (pre != null && tmp != null && tmp.equals(pre))
                st.pop();
            else
                st.push(ch);
        }
        if (st.isEmpty())
            return true;
        else
            return false;
    }
}

File: 20.valid-parentheses.cs
/*
 * @lc app=leetcode id=20 lang=csharp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
public class Solution
{
    public bool IsValid(string s) {
        var map = new Dictionary<char, char> { { ')', '(' }, { ']', '[' }, { '}', '{' } };
        var stack = new Stack<char>();
        foreach(var item in s) {
            if(item == '(' || item == '[' || item == '{')
                stack.Push(item);
            if(map.ContainsKey(item)) {
                var ch = map[item];
                if(stack.Count != 0 && stack.Peek() == ch)
                    stack.Pop();
                else
                    return false;
            }
        }
        if(stack.Count != 0)
            return false;
        return true;
    }
}
// @lc code=end

